### ISSUE TITLE
Cannot display unicode characters on Windows

### ISSUE THREAD

**Issue by gyk** [babashka/babashka#1009]
**version**

v0.6.0

**platform**

Windows 11

**problem**

Babashka can only print ASCII characters in Windows Terminal.

**repro**

Execute either in REPL or from a script (encoding: UTF-8):

```clojure
(do
  (prn "™")
  (prn "λ")
  (prn "⚙️")
  (prn "中文"))
```

Output:

```
"?

"?"
"??"
"??"
```

<br/>

I also tried [this example](https://github.com/babashka/babashka/issues/359) using PowerShell 7:
```sh
PS > bb -e '(prn "bépo àê")'
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  EOF while reading, expected ) to match ( at [1,1]
Data:     {:type :edamame/error, :line 1, :column 10, :edamame/expected-delimiter ")", :edamame/opened-delimiter "(", :edamame/opened-delimiter-loc {:row 1, :col 1}}
Location: <expr>:1:10
Phase:    parse

----- Context ------------------------------------------------------------------
1: (prn bépo
            ^--- EOF while reading, expected ) to match ( at [1,1]
```

However, the command above does work in Cygwin:

```sh
$ bb -e '(prn "bépo àê")'
"bépo àê"
```

**Comment by borkdude** [comment]
Can you make a repro using Clojure in Windows? That is the reference implementation. Just to make sure this is not a problem of shells instead of bb.

**Comment by gyk** [comment]
Yes, Clojure (installed with Scoop) behaves exactly the same.

(My bad, I usually run Clojure in WSL 2...)

**Comment by borkdude** [comment]
So this issue can be closed, or is there something you expect?

**Comment by gyk** [comment]
I think so, and passing `-Dfile.encoding=UTF-8` might solve the problem. Thanks!

**Comment by robinlahtinen** [comment]
I'm having problems with UTF-8 on Windows. Would changing the Babashka build options to force UTF-8 solve the problem similar to [this issue](https://github.com/NixOS/nixpkgs/pull/153457) or passing `-Dfile.encoding=UTF-8`?

Babashka:
```
bb -e '(prn "bépo àê")'
"b�po ��"
```

Clojure REPL:
```
(prn "bépo àê")
"bépo àê"
```

**Comment by borkdude** [comment]
What happens if _you_ pass `-Dfile.encoding=UTF-8` to bb?

**Comment by robinlahtinen** [comment]
There is no difference if you mean it like this:

```
bb -e '(prn "bépo àê")' -Dfile.encoding=UTF-8
"b�po ��"
```

**Comment by borkdude** [comment]
You have to pass `-D` as the very first argument

**Comment by robinlahtinen** [comment]
Result:
```
bb -Dfile.encoding=UTF-8 -e '(prn "bépo àê")'
----- Error --------------------------------------------------------------------
Type:     java.lang.Exception
Message:  File does not exist: -Dfile
```

**Comment by borkdude** [comment]
What version of bb is this? You should be able to pass -D arguments. I'm getting this:

```
$ bb -Dfile.encoding=UTF-8 -e '(prn "bépo àê")'
"bépo àê"
```

**Comment by robinlahtinen** [comment]
Using babashka v1.3.187-SNAPSHOT on PowerShell 7. Snapshot is the current build from AppVeyor, but the latest stable also has the same problem.

PowerShell 7 shell, with the new prompt:
```
bb "-Dfile.encoding=UTF-8" -e '(prn "bépo àê")'
"b�po ��"
```

PowerShell 7 shell:
```
bb -Dfile.encoding=UTF-8 -e '(prn "bépo àê")'
----- Error --------------------------------------------------------------------
Type:     java.lang.Exception
Message:  File does not exist: -Dfile
```

cmd.exe shell:
```
bb -Dfile.encoding=UTF-8 -e '(prn "bépo àê")'
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  EOF while reading, expected ) to match ( at [1,2]
Data:     {:type :edamame/error, :line 1, :column 6, :edamame/expected-delimiter ")", :edamame/opened-delimiter "(", :edamame/opened-delimiter-loc {:row 1, :col 2}}
Location: <expr>:1:6
Phase:    parse

----- Context ------------------------------------------------------------------
1: '(prn
        ^--- EOF while reading, expected ) to match ( at [1,2]

----- Stack trace --------------------------------------------------------------
edamame.impl.parser/throw-reader       - <built-in>
edamame.impl.parser/parse-to-delimiter - <built-in>
edamame.impl.parser/parse-list         - <built-in>
edamame.impl.parser/dispatch           - <built-in>
edamame.impl.parser/parse-next         - <built-in>
edamame.impl.parser/dispatch           - <built-in>
edamame.impl.parser/parse-next         - <built-in>
edamame.core/parse-next                - <built-in>
```

**Comment by borkdude** [comment]
Please don't conflate shell escaping issues with babashka issues. cmd.exe does not support `'...'` escaping. Like I said you need to completely quote the `"-Dfile.encoding=UTF-8"` argument in powershell (second example in powershell 7).
This doesn't solve the issue though. Perhaps you can try compiling bb.exe locally to see if that gives the same issue?

**Comment by borkdude** [comment]
There is some info here too:
https://stackoverflow.com/questions/54952/java-utf-8-and-windows-console

**Comment by borkdude** [comment]
This is what I'm getting with clojure:

```
clj -M -Dfile.encoding=UTF-8 -e  """bépo àê"""
"b�po ��"
```

I think it might be best to first try to reproduce this issue with just clojure/JVM and then see if we can fix it in bb.

**Comment by robinlahtinen** [comment]
Using Windows-1252 encoding seems to work, but UTF-8 does not.

PowerShell with code page 65001 (UTF-8):
```
chcp 65001 && bb "-Dfile.encoding=UTF-8" -e '(prn "bépo àê")'
Active code page: 65001
"b�po ��"
```

PowerShell with code page 1252 ("ANSI", [Windows-1252](https://en.wikipedia.org/wiki/Windows-1252)):
```
chcp 1252 && bb "-Dfile.encoding=UTF-8" -e '(prn "bépo àê")'
Active code page: 1252
"bépo àê"
```

**Comment by Hamido-san** [comment]
any chance we can get an actual fix 5 years into reporting this issue?

**Comment by borkdude** [comment]
@Hamido-san Any chance you can read through this issue to see what's actually going on before making entitled remarks? If you read this issue carefully you can see that the conclusion was that this is not a bb issue since the behavior is the same with JVM Clojure. If you think otherwise, feel free to make a new issue with a clear diagnosis of what's going on and that there is a difference with JVM Clojure. Thank you.

**Comment by Hamido-san** [comment]
no no, sorry for the outburst. I don't know what came over me. can't even diagnose a JVM / GraalVM issue on my own. heck, this wouldn't be the Clojure ecosystem if we treated supporting the entitled windows users as anything more than an afterthought. my bad.

**Comment by borkdude** [comment]
@Hamido-san "entitled" doesn't have anything to do with Windows users. Babashka supports Windows fully, period. It was about someone coming in here as their first interaction on this Github repo saying "just fix this, it took five years, what's keeping so long" without actually understanding if there's even something to be fixed. If you can clarify this in a clear way that is compatible with the [Clojure etiquette](https://clojure.org/community/etiquette) that's not just a repetition of what's already been discussed in this issue, I'd be more than willing to help.

**Comment by borkdude** [comment]
I got something working on Windows in Powershell now. Consider this file:

``` clojure
(prn :> (System/getProperty "file.encoding"))

(alter-var-root #'*out* (constantly (new java.io.OutputStreamWriter System/out)))

(do
  (prn "™")
  (prn "λ")
  (.println System/out "⚙️")
  (prn "中文"))

(binding [*out* *out* #_(java.io.OutputStreamWriter. System/out #_(java.nio.charset.Charset/forName "UTF-8"))]
  (println "✓ works")
  (prn "⚙️"))
```

- Since GraalVM 23 (JDK 18+)  the file.encoding is UTF-8 by default. This is verified by printing the system property at the top of the script. This works.
- You need to set `$OutputEncoding = [Console]::OutputEncoding = [System.Text.UTF8Encoding]::new()` in Powershell to fix the output to UTF-8. You can possibly do this in a profile. See https://stackoverflow.com/a/57134096
- I need to reset `*out*` with `(alter-var-root #'*out* (constantly (new java.io.OutputStreamWriter System/out)))`. I don't know why. Perhaps because clojure is build-time initialized, `System/out` has the wrong encoding or so. But when I reset it, it works.
- The last form demonstrates the same thing (but optionally with an explicit encoding)

So in summary: 

- Windows users need to set some stuff for Powershell to work with UTF-8 correctly and/or need to opt-in on the beta feature (see SO link).
- The `*out*` variable needs to be reset at runtime. This might have to do with a bug in native-image and/or an issue with built-time initialization. I can possibly fix this in babashka by resetting out myself at runtime + investigate with a repro.

<img width="692" height="368" alt="Image" src="https://github.com/user-attachments/assets/660675a6-992e-44e6-a76d-d72f81777b7f" />

**Comment by borkdude** [comment]
After trying several things, the above fix is the only one that consistently works for me when building on Github Actions.

https://github.com/babashka/babashka/commit/03e0e9a19a6ed6c6e0231a4f7ed06082d6d53d04

You can download this binary which has this "fix":

https://github.com/babashka/babashka/actions/runs/18006768113/artifacts/4103894502

You still need to set `$OutputEncoding = [Console]::OutputEncoding = [System.Text.UTF8Encoding]::new()` in your powershell (you can do so in a profile instead of manually).

If some people can test this, I'll merge the fix/workaround for Windows.

**Comment by borkdude** [comment]
(What's bothering me about this is that I can't reproduce the problem with a simple Java class and native-image like here: https://github.com/borkdude/graal-repros/tree/init-at-build-time-output-writer-repro)

**Comment by borkdude** [comment]
Made a GraalVM issue here: https://github.com/oracle/graal/issues/12249

**Comment by borkdude** [comment]
@Hamido-san The issue is fixed now (using a workaround for Windows and I filed a GraalVM issue). You can try a development version of babashka if you like or wait until the next release which should be soon-ish. I came across a bit harsh to you out of frustration but I'm open to the possibility that I misinterpreted your initial post and probably should have reacted differently. With that out of the way, hopefully the experience in Windows powershell is better now.
